## Question -1
ans - 

1. Ordinal Encoding:

* .Definition: Ordinal encoding is used when there is a clear, meaningful, and ordered relationship among the categories of a categorical variable. This means that the categories have a specific rank or hierarchy.

* .Example: Consider a dataset with an "Education Level" feature, where the categories are "High School," "Associate's Degree," "Bachelor's Degree," "Master's Degree," and "Ph.D." These categories have an inherent order, with "Ph.D." being higher in rank than "Bachelor's Degree," and so on.

* .Encoding: In ordinal encoding, you assign unique integer values to the categories based on their order. For example, you might assign 1 to "High School," 2 to "Associate's Degree," and so on.

* .Use Case: Ordinal encoding is appropriate when there's a meaningful and interpretable ordinal relationship between categories. It's often used in cases where there is a clear hierarchy, such as education levels, customer satisfaction ratings, or age groups.

2. Label Encoding:

* .Definition: Label encoding is used when there is no inherent order or hierarchy among the categories of a categorical variable. In label encoding, you assign unique integer values to each category for the purpose of distinction, but the order of these values does not carry any meaning.

* .Example: Consider a dataset with a "Color" feature, where the categories are "Red," "Green," and "Blue." These colors have no intrinsic order or ranking.

* .Encoding: In label encoding, you simply assign unique integers to the categories without implying any order. For example, you might assign 1 to "Red," 2 to "Green," and 3 to "Blue."

* .Use Case: Label encoding is suitable when the categories have no inherent order, and you want to represent them numerically. It is commonly used for features like gender, color, or other non-ordinal categories.



When to Choose One Over the Other:

>> . Choose Ordinal Encoding when:

* . The categories have a clear and meaningful order or hierarchy.
* .The order of the categories is important for the problem or analysis.
* . You can justify an ordinal relationship between the categories.

--. Example: When encoding education levels, as the level of education increases, the value assigned should reflect this increase in rank.

>> .Choose Label Encoding when:

* .The categories have no inherent order or ranking, and their order doesn't carry any meaningful information.
* .You need a simple way to represent non-ordinal categories numerically.

--. Example: When encoding colors, there is no intrinsic order among colors, so label encoding is appropriate.

## Question - 2
ans - 

Here's how Target Guided Ordinal Encoding works:

1. Calculate the Mean of the Target Variable for Each Category: For each category in the categorical variable, calculate the mean of the target variable within that category. This means you are determining the probability or likelihood of the target variable taking a certain value within each category.

2. Order Categories Based on the Calculated Means: Sort the categories based on their associated mean values in ascending or descending order. If the target variable represents a binary outcome (e.g., 0 or 1), you can sort the categories from the category with the lowest mean to the category with the highest mean. If the target variable has multiple classes, you can sort based on ascending or descending means.

3. Assign Ordinal Values: Assign ordinal values (integers) to the sorted categories. You can assign values starting from 1, 2, 3, and so on, according to their order in the sorted list.

4. Use the Encoded Categories: Replace the original categorical variable with the ordinal values assigned in the previous step.


>>. Here's an example of when you might use Target Guided Ordinal Encoding in a machine learning project:

* . Scenario: You are working on a customer churn prediction project for a telecom company. One of the categorical features is "Contract Type," which has three categories: "Month-to-Month," "One Year," and "Two Year." In this case, there is a natural order to the contract types, with "Month-to-Month" being the most flexible and "Two Year" being the most committed.

--.  can use Target Guided Ordinal Encoding as follows:

1. Calculate the mean churn rate (the likelihood of a customer churning) for each of the three contract types. You analyze how many customers of each contract type have churned (target = 1) and how many have not churned (target = 0).

2. Sort the contract types based on their mean churn rates, from lowest to highest.

3. Assign ordinal values to the sorted contract types. For example, you can assign 1 to "Month-to-Month," 2 to "One Year," and 3 to "Two Year."

4. Replace the original "Contract Type" categorical variable with the ordinal values.

## Question - 3
ans - 

Covariance is a statistical measure that quantifies the degree to which two random variables (or data columns) change together. It describes the joint variability of two variables and helps assess whether they tend to increase or decrease at the same time. In other words, it indicates the direction of the linear relationship between the two variables.

* . The importance of covariance in statistical analysis can be summarized as follows:

1. Measuring Association: Covariance provides a numerical value that indicates whether two variables are positively or negatively associated. A positive covariance suggests that when one variable increases, the other tends to increase as well, while a negative covariance indicates that they move in opposite directions.

2. Assessing Relationships: In various fields, such as finance, economics, and social sciences, understanding the relationships between variables is crucial. Covariance helps in assessing the strength and direction of these relationships.

3. Portfolio Diversification: In finance, covariance is used to assess the relationship between the returns of different assets in a portfolio. A low or negative covariance between assets can be used for diversification to reduce risk.

4. Data Preprocessing: In data preprocessing for machine learning, covariance can help identify redundant features or variables. Features with high positive covariance may contain similar information, while those with high negative covariance may provide complementary information.

--. Covariance is calculated using the following formula:

For two variables X and Y with n data points:

Cov(X, Y) = Σ (Xi - μX) * (Yi - μY) / n

Where:

Xi and Yi are individual data points for X and Y, respectively.

μX and μY are the means (averages) of variables X and Y, respectively.

n is the total number of data points.

## Question  - 4
ans - 

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
import numpy as np



In [11]:
df = pd.DataFrame({'color':['red', 'green', 'blue'], 
                  'size':['small', 'medium', 'large'], 
                  'material': ['wood', 'metal', 'plastic']})


df

color    size material
0    red   small     wood
1  green  medium    metal
2   blue   large  plastic

In [14]:
encoder = LabelEncoder()

encoded = encoder.fit_transform(df['color'])

df['encoded_color'] = pd.DataFrame(encoded)


df['encoded_size'] = pd.DataFrame(encoder.fit_transform(df['size']))

df['encoded_material'] = pd.DataFrame(encoder.fit_transform(df['material']))

df


color    size material  encoded_color  encoded_size  encoded_material
0    red   small     wood              2             2                 2
1  green  medium    metal              1             1                 0
2   blue   large  plastic              0             0                 1

## Question - 5
ans - 

In [17]:
import pandas as pd

df = pd.DataFrame({'age':[25, 30, 35, 40, 45],
                  'income':[50000, 60000, 75000, 80000, 90000],
                  'education':[12, 16, 18, 14, 20]})

df

age  income  education
0   25   50000         12
1   30   60000         16
2   35   75000         18
3   40   80000         14
4   45   90000         20

In [18]:
df.cov()

age       income  education
age            62.5     125000.0       17.5
income     125000.0  255000000.0    37500.0
education      17.5      37500.0       10.0

* .Interpretation:

1. Age Variance: The variance of Age is 62.5. This indicates that the Age values in the dataset have significant variability or spread around the mean Age.

2. Income Variance: The variance of Income is 255,000,000.0. This shows a high degree of variability in Income values, indicating that Income varies widely in the dataset.

3. Education Level Variance: The variance of Education Level is 10.0. This represents the spread or variability in Education Level values, which is relatively low compared to Age and Income.

4. Age-Income Covariance: The covariance between Age and Income is 125,000.0. This positive covariance suggests that, on average, as Age increases, Income tends to increase. However, the magnitude of covariance does not provide information about the strength of this relationship, so it's not possible to determine how strong the relationship is based on the covariance alone.

5. Age-Education Level Covariance: The covariance between Age and Education Level is 17.5. This positive covariance indicates a positive relationship between Age and Education Level, but it doesn't provide information about the strength of the relationship.

6. Income-Education Level Covariance: The covariance between Income and Education Level is 37,500.0. This positive covariance suggests that, on average, as Income increases, Education Level tends to increase. However, the magnitude alone doesn't provide information about the strength of the relationship.

## Question - 6
ans- 

1. Gender (Binary Categorical Variable: Male/Female):

* .Encoding Method: Label Encoding. This variable has two categories, "Male" and "Female," which are binary and do not have an inherent ordinal relationship. Label encoding is suitable, with 0 representing one category and 1 representing the other.

2. Education Level (Ordinal Categorical Variable: High School/Bachelor's/Master's/PhD):

* .Encoding Method: Ordinal Encoding. Education levels have a clear ordinal relationship, with "High School" < "Bachelor's" < "Master's" < "PhD." Ordinal encoding should be used to preserve this order, with integer values assigned based on the hierarchy.

3. Employment Status (Ordinal Categorical Variable: Unemployed/Part-Time/Full-Time):

* .Encoding Method: Ordinal Encoding. Employment status also has a clear order or hierarchy, with "Unemployed" < "Part-Time" < "Full-Time." Ordinal encoding should be used to capture this ordinal information with appropriate integer values.


--. 

* ."Gender" should be encoded using Label Encoding.
* ."Education Level" should be encoded using Ordinal Encoding.
* ."Employment Status" should also be encoded using Ordinal Encoding.

## Question - 7
ans - 

In [1]:
import pandas as pd

data = {'Temperature': [25.0, 28.0, 22.0, 30.0, 26.0],
       'Humidity': [60, 55, 75, 45, 70],
       'Weather Condition': ['Sunny', 'Cloudy', 'Rainy', 'Sunny', 'Rainy'],
       'Wind Direction': ['North', 'South', 'East', 'West', 'North']}


df = pd.DataFrame(data)
df

Temperature  Humidity Weather Condition Wind Direction
0         25.0        60             Sunny          North
1         28.0        55            Cloudy          South
2         22.0        75             Rainy           East
3         30.0        45             Sunny           West
4         26.0        70             Rainy          North

In [7]:
from sklearn.preprocessing import LabelEncoder

label_encoders = {}
encoded_data = df.copy()

encoder = LabelEncoder()

for col in df.select_dtypes(include = 'object'):
    encoded_data[col] = encoder.fit_transform(df[col])
    
    
    
    
    

encoded_data
    
    
    

Temperature  Humidity  Weather Condition  Wind Direction
0         25.0        60                  2               1
1         28.0        55                  0               2
2         22.0        75                  1               0
3         30.0        45                  2               3
4         26.0        70                  1               1

In [8]:
covariance = encoded_data.cov()
covariance

Temperature  Humidity  Weather Condition  Wind Direction
Temperature               9.20    -32.75               0.20            3.40
Humidity                -32.75    142.50              -2.75          -13.00
Weather Condition         0.20     -2.75               0.70            0.15
Wind Direction            3.40    -13.00               0.15            1.30

interpretation -:
    
    
1. Temperature and Humidity:

* .Covariance: -32.75

* .Negative covariance indicates that as "Temperature" increases, "Humidity" tends to decrease, and as "Temperature" decreases, "Humidity" tends to increase. The magnitude of -32.75 indicates a strong negative relationship between temperature and humidity.


2. Temperature and Weather Condition (Encoded):

* .Covariance with "Weather Condition" variables: These values indicate the relationships between "Temperature" and different weather conditions.
* .The positive covariance with "Weather_Sunny" (0.20) suggests that higher temperatures are associated with a slightly increased likelihood of sunny weather. However, the magnitude of the covariances is relatively low, indicating weak relationships between "Temperature" and weather conditions.


3. Temperature and Wind Direction (Encoded):

* .Covariance with "Wind Direction" variables: These values indicate the relationships between "Temperature" and different wind directions.
* .Positive covariances suggest that as "Temperature" increases, certain wind directions may become more likely. The positive covariance with "Wind Direction" variables indicates moderate associations, but the values do not provide strong evidence of the relationships' strength.


4. Humidity and Weather Condition (Encoded):

* .Covariance with "Weather Condition" variables: These values indicate the relationships between "Humidity" and different weather conditions.

* .The negative covariance with "Weather_Rainy" (-2.75) suggests that higher humidity is associated with a reduced likelihood of rainy weather. The magnitude of these covariances indicates moderate relationships between "Humidity" and weather conditions.

5. Humidity and Wind Direction (Encoded):

* .Covariance with "Wind Direction" variables: These values indicate the relationships between "Humidity" and different wind directions.

* .Negative covariances suggest that as "Humidity" increases, certain wind directions may become less likely. The magnitude of these covariances indicates moderate associations between "Humidity" and wind directions.